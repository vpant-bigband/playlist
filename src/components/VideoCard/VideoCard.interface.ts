export interface Image {
  url: string;
  width: number;
  height: number;
}

interface Thumbnails {
  default: Image;
  medium: Image;
  high: Image;
}

export interface VideoCardProps {
  channelId: string;
  title: string;
  description: string;
  thumbnails: Thumbnails;
  publishedAt: string;
  channelTitle: string;
  liveBroadcastContent: string;
  publishTime: string;
}