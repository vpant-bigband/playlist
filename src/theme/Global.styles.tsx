import React from 'react';
import { Global, css } from '@emotion/react';
import theme from './theme.constant';

const GlobalStyle: React.FunctionComponent = () => (
  <Global
    styles={css`
      body {
        font-family: ${theme.fonts.sansSerif};
        font-size: ${theme.fontSizes[2]}px;
        line-height: ${theme.lineHeights.title};
        margin: 0px;
      }
      main {
        max-width: 1440px;
        margin-left: auto;
        margin-right: auto;
      }
    `}
  />
);

export default GlobalStyle;
