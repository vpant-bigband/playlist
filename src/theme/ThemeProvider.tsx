import React from 'react';
import { ThemeProvider } from '@emotion/react';
import ThemeConstant, { ThemeType } from './theme.constant';
import GlobalStyle from './Global.styles';

interface ThemeProps {
  children: React.ReactChildren | React.ReactElement | React.ReactElement[];
};

const Theme: React.FunctionComponent<ThemeProps> = ({ children }) => {
  return (
    <ThemeProvider theme={ThemeConstant as ThemeType}>
      <GlobalStyle />
      {children}
    </ThemeProvider>
  );
}



export default Theme;