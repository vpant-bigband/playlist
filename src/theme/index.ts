export { default as Theme } from './ThemeProvider';