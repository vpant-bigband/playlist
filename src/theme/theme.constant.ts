const theme = {
  breakpoints: ['576px', '767px', '992px', '1200px'],
  BREAKPOINTS: [576, 767, 992, 1200],
  space: [0, 4, 8, 16, 24, 32, 64, 128, 256, 512],
  fontSizes: [10, 12, 14, 16, 18, 20, 24, 36, 48, 80, 96],
  fontWeights: [100, 200, 300, 400, 500, 600, 700, 800, 900],
  lineHeights: {
    solid: 1,
    title: 1.25,
    copy: 1.5,
  },
  fonts: {
    serif: 'athelas, georgia, times, serif',
    sansSerif: '"Roboto", sans-serif',
    montserrat: 'Montserrat',
  },
  borders: [0, '1px solid', '2px solid', '4px solid', '8px solid', '16px solid', '32px solid'],
  width: [16, 24, 32, 64, 128, 256],
  heights: [16, 24, 32, 64, 128, 256],
  maxWidths: [16, 32, 64, 128, 256, 512, 768, 1024, 1536],
  colors: {
    // Copy Typography
    blue: '#0E5AA7',
    darkBlue: '#0a437c',
    purple: '#6F1C60',
    lightBlue: '#b7cee5',
    deepRed: '#8A0C0C',
    // Greys
    ashGray: '#F9F7F7',
    lightGray: '#EAEAEA',
    darkGray: '#2D2D2D',

    white: '#fff',
    transparent: 'transparent',
    black: '#000',
    // Notification color
    error: '#FEF4F4',
    warning: '#FCF6E7',
    information: '#ECF3FA',
    success: '#F1FAF9',
  },
};

export type ThemeType = typeof theme;

export default theme;