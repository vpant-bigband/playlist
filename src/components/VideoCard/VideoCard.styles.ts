import styled from '@emotion/styled';
import theme, { ThemeType } from '../../theme/theme.constant';
import { variant, width, FontSizeProps, fontSize, typography, TypographyProps, padding, PaddingProps, color, ColorProps, display, DisplayProps } from 'styled-system';


export const VideoCardWrapper = styled.div<{ theme?: ThemeType }>`
  display: flex;
  padding: 20px;
  border-bottom: ${(props) => theme.borders[1] + ' ' + theme.colors.lightBlue}; 
`;

export const VideoCardImage = styled.img<{ resolution: string; height: number; width: number}>`
  ${({ height, width }) => ({
    height,
    width,
  })}
  ${variant({
    prop: 'resolution',
    variants: {
      'high': {
        padding: 10,
      },
      'medium': {
        padding: 5,
      },
      'low': {
        padding: 15,
        marginTop: 'auto',
        marginBottom: 'auto',
      },
    },
  })}
`;

export const VideoCardContent = styled.div`
  display: flex;
  flex-direction: column;
  padding: 10px;
  width: 50%;
`;

export const Text = styled.span<FontSizeProps & PaddingProps & ColorProps & DisplayProps>`
  ${fontSize}
  ${color}
  ${padding}
  ${display}
  word-break: break-word;
`;

export const VideoCardTitle = styled(Text)<{ theme?: ThemeType } & PaddingProps>`
  ${padding}
  font-weight: ${({theme}) => theme.fontWeights[5]};
  padding-top: 10px;
`;

export const VideoCardDescription = styled(Text)<{ theme?: ThemeType }>`
  font-weight: ${({theme}) => theme.fontWeights[4]};
  width: 85%;
`;
