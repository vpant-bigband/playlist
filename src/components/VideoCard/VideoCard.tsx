import React, { useMemo } from 'react';
import { VideoCardProps, Image } from './VideoCard.interface';
import { VideoCardWrapper, VideoCardImage, VideoCardContent, VideoCardTitle, VideoCardDescription, Text } from './VideoCard.styles';
import theme from '../../theme/theme.constant';
import useBreakpoints from '../utils/useBreakpoints.hooks';

const VideoCard: React.FunctionComponent<VideoCardProps> = ({ title, thumbnails, description, channelTitle}) => {
  const device = useBreakpoints();
  const [{ height, url, width }, resolution] = useMemo(() => {
    const { isMobile, isPad1 } = device;
    let thumbnail = [thumbnails.high, 'high'];
    if (isMobile) {
      thumbnail = [thumbnails.default, 'low'];
    } 
    if (isPad1) {
      thumbnail = [thumbnails.medium, 'medium'];
    }
    return thumbnail;
  }, [thumbnails, device]) as [Image, string];

  return (
    <VideoCardWrapper data-testid='video-card-wrapper'>
      <VideoCardImage data-testid='video-card-image' src={url} resolution={resolution} height={height} width={width} />
      <VideoCardContent>
        <VideoCardTitle data-testid='video-card-title' pb={[0, 20]} fontSize={[1, 2, 4]} dangerouslySetInnerHTML={{ __html: title }} />
        <Text display={['none', 'none', 'flex']} data-testid='video-card-channel-title' paddingBottom="10px" color={theme.colors.blue} fontSize={[1,2]}>{channelTitle}</Text>
        <VideoCardDescription display={['none', 'none', 'flex']} data-testid='video-card-description' fontSize={[1, 2, 3]}>
          {description}
        </VideoCardDescription>
      </VideoCardContent>
    </VideoCardWrapper>
  );
}

export default VideoCard;