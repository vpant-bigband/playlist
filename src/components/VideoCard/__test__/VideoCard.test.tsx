import React from 'react';
import { render } from '@testing-library/react';
import VideoCard from '../VideoCard';
import VideoCardMock from '../__mocks__/VideoCard.mock';
import { Theme } from '../../../theme';


const renderSetup = () => render(<Theme><VideoCard {...VideoCardMock} /></Theme>);
describe('VideoCard Component test suite', () => {
  it('should match snapshot', () => {
      const { container } = renderSetup();
      expect(container).toMatchSnapshot();
  });

  it('should render image', () => {
    const { getByTestId } = renderSetup();
    const image = getByTestId('video-card-image');
    expect(image).toBeInTheDocument();
    expect(image).toHaveProperty('src', VideoCardMock.thumbnails.high.url);
  });
  it('should render with title', () => {
    const { getByTestId } = renderSetup();
    const title = getByTestId('video-card-title');
    expect(title).toBeInTheDocument();
    expect(title).toHaveTextContent(VideoCardMock.title);
  });
  it('should render with description', () => {
    const { getByTestId } = renderSetup();
    const title = getByTestId('video-card-description');
    expect(title).toBeInTheDocument();
    expect(title).toHaveTextContent(VideoCardMock.description);
  });

  it('should render with channel title', () => {
    const { getByTestId } = renderSetup();
    const title = getByTestId('video-card-channel-title');
    expect(title).toBeInTheDocument();
    expect(title).toHaveTextContent(VideoCardMock.channelTitle);
  });
});