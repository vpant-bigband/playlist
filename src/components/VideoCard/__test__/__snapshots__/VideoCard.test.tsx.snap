// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard Component test suite should match snapshot 1`] = `
<div>
  <div
    class="css-bj2njj"
    data-testid="video-card-wrapper"
  >
    <img
      class="css-1u6w3ei"
      data-testid="video-card-image"
      height="360"
      src="https://i.ytimg.com/vi/mFoNq5LHaNg/hqdefault_live.jpg"
      width="480"
    />
    <div
      class="css-ng8x8e"
    >
      <span
        class="css-15u1n5f"
        data-testid="video-card-title"
        font-size="1,2,4"
      >
        New Hindi Song 2020 December ðŸ’– Top Bollywood Romantic Love Songs 2020 ðŸ’– Best Indian Songs 2020
      </span>
      <span
        class="css-5zodk7"
        color="#0E5AA7"
        data-testid="video-card-channel-title"
        display="none,none,flex"
        font-size="1,2"
      >
        Bollywood Hits Songs
      </span>
      <span
        class="css-xb8gpa"
        data-testid="video-card-description"
        display="none,none,flex"
        font-size="1,2,3"
      >
        New Hindi Song 2020 December Top Bollywood Romantic Love Songs 2020 Best Indian Songs 2020 ...
      </span>
    </div>
  </div>
</div>
`;
